src/bugpoint.rs

```diff
@@ -0,0 +1,312 @@
+                    let results = func.dfg.inst_results(prev_inst);
+                    if results.len() == 1 {
+                        let ty = func.dfg.value_type(results[0]);
+                        func.dfg.replace(prev_inst).iconst(ty, 0);
+                        PhaseStepResult::Replaced(format!(
+                            "Replace inst {} with iconst.{}",
+                            prev_inst, ty
+                        ))
+                    } else {
+                        PhaseStepResult::NoChange
+                    }
+                } else {
+                    *self = Phase::ReplaceInstWithTrap(
+                        first_ebb,
+                        func.layout.first_inst(first_ebb).unwrap(),
+                    );
+                    PhaseStepResult::NextPhase("replace inst with trap", inst_count(func))
+                }
+            }
+            Phase::ReplaceInstWithTrap(ref mut ebb, ref mut inst) => {
+                if let Some((_prev_ebb, prev_inst)) = next_inst_ret_prev(func, ebb, inst) {
+                    func.dfg.replace(prev_inst).trap(TrapCode::User(0));
+                    PhaseStepResult::Replaced(format!("Replace inst {} with trap", prev_inst))
+                } else {
+                    *self = Phase::RemoveEbb(first_ebb);
+                    PhaseStepResult::NextPhase("remove ebb", ebb_count(func))
+                }
+            }
+            Phase::RemoveEbb(ref mut ebb) => {
+                let prev_ebb = *ebb;
+                if let Some(next_ebb) = func.layout.next_ebb(*ebb) {
+                    *ebb = next_ebb;
+                    func.layout.remove_ebb(*ebb);
+                    PhaseStepResult::Shrinked(format!("Remove ebb {}", prev_ebb))
+                } else {
+                    PhaseStepResult::Finished
+                }
+            }
+        }
+    }
+}
+
+fn next_inst_ret_prev(func: &Function, ebb: &mut Ebb, inst: &mut Inst) -> Option<(Ebb, Inst)> {
+    let prev = (*ebb, *inst);
+    if let Some(next_inst) = func.layout.next_inst(*inst) {
+        *inst = next_inst;
+        return Some(prev);
+    } else if let Some(next_ebb) = func.layout.next_ebb(*ebb) {
+        *ebb = next_ebb;
+        *inst = func.layout.first_inst(*ebb).expect("no inst");
+        return Some(prev);
+    } else {
+        return None;
+    }
+}
+
+fn ebb_count(func: &Function) -> usize {
+    func.layout.ebbs().count()
+}
+
+fn inst_count(func: &Function) -> usize {
+    func.layout
+        .ebbs()
+        .map(|ebb| func.layout.ebb_insts(ebb).count())
+        .sum()
+}
+
+fn resolve_aliases(func: &mut Function) {
+    for ebb in func.layout.ebbs() {
+        for inst in func.layout.ebb_insts(ebb) {
+            func.dfg.resolve_aliases_in_arguments(inst);
+        }
+    }
+}
+
+fn reduce(isa: &TargetIsa, mut func: Function, verbose: bool) {
+    let (orig_ebb_count, orig_inst_count) = (ebb_count(&func), inst_count(&func));
```

Comment:
We probably don't want to reduce a function which compilation doesn't panic in the first place, right?

Reply:
Yes